core/src/translate/param.rs

```diff
@@ -0,0 +1,529 @@
+use {
+    super::TranslateError,
+    crate::{
+        ast::{AstLiteral, DataType, DateTimeField, Expr},
+        data::{Interval, Point},
+    },
+    bigdecimal::BigDecimal,
+    chrono::{NaiveDate, NaiveDateTime, NaiveTime, Timelike},
+    rust_decimal::Decimal,
+    std::net::IpAddr,
+    uuid::Uuid,
+};
+
+#[derive(Debug, Clone)]
+pub enum ParamLiteral {
+    Literal(AstLiteral),
+    TypedString {
+        data_type: DataType,
+        value: String,
+    },
+    Interval {
+        expr: Box<ParamLiteral>,
+        leading_field: Option<DateTimeField>,
+        last_field: Option<DateTimeField>,
+    },
+}
+
+impl ParamLiteral {
+    pub const fn null() -> Self {
+        Self::Literal(AstLiteral::Null)
+    }
+
+    pub fn into_expr(self) -> Expr {
+        match self {
+            ParamLiteral::Literal(literal) => Expr::Literal(literal),
+            ParamLiteral::TypedString { data_type, value } => {
+                Expr::TypedString { data_type, value }
+            }
+            ParamLiteral::Interval {
+                expr,
+                leading_field,
+                last_field,
+            } => Expr::Interval {
+                expr: Box::new(expr.into_expr()),
+                leading_field,
+                last_field,
+            },
+        }
+    }
+}
+
+fn into_number_literal(value: impl ToString) -> Result<AstLiteral, TranslateError> {
+    let value_str = value.to_string();
+    match value_str.parse::<BigDecimal>() {
+        Ok(number) => Ok(AstLiteral::Number(number)),
+        Err(_) => Err(TranslateError::InvalidParamLiteral { value: value_str }),
+    }
+}
+
+pub trait IntoParamLiteral {
+    fn into_param_literal(self) -> Result<ParamLiteral, TranslateError>;
```

Comment:
Could you please add a brief doc comment here that includes a # Errors section explaining when `into_param_literal` may return a TranslateError? For example:

```rust
/// Converts the value into a [`ParamLiteral`] so it can be bound as a query parameter.
///
/// # Errors
///
/// Returns an error if the value cannot be represented as a [`ParamLiteral`],
/// for example when a numeric conversion or validation fails, or when downstream
/// constraints such as the `NonZeroUsize` placeholder index requirement reject it.
fn into_param_literal(self) -> Result<ParamLiteral, TranslateError>;
```